File: dcp-solutions/2024-12-06/solution.py
```python
def validate_rules(rules):
    directions = {
        'N': (0, 1),
        'S': (0, -1),
        'E': (1, 0),
        'W': (-1, 0)
    }

    pos_a = {'x': 0, 'y': 0}
    pos_b = {}

    for rule in rules:
        loc_a, d_a, loc_b = rule.split()
        dx, dy = directions[d_a]

        if d_a == 'N':
            pos_b['x'] += dx
            pos_b['y'] += dy
        elif d_a == 'S':
            pos_b['x'] -= dx
            pos_b['y'] -= dy
        elif d_a == 'E':
            pos_b['x'] += dx
            pos_b['y'] += dy
        elif d_a == 'W':
            pos_b['x'] -= dx
            pos_b['y'] -= dy

        if loc_b not in pos_b:
            pos_b[loc_b] = {'x': 0, 'y': 0}

        x_diff = pos_a['x'] - pos_b[loc_b]['x']
        y_diff = pos_a['y'] - pos_b[loc_b]['y']

        if (d_a == 'N' and x_diff != 0) or \
           (d_a == 'S' and x_diff == 0) or \
           (d_a == 'E' and y_diff != 0) or \
           (d_a == 'W' and y_diff == 0):
            return False

        pos_a = pos_b[loc_b]

    return True
```

File: dcp-solutions/2024-12-06/solution.rb
```ruby
def validate_rules(rules)
  # Mapping of directions to their opposite directions
  direction_map = {
    'N' => 'S',
    'S' => 'N',
    'E' => 'W',
    'W' => 'E'
  }

  # Mapping of direction pairs to their corresponding directions
  pair_map = {
    'NE' => :north_east,
    'NW' => :north_west,
    'SE' => :south_east,
    'SW' => :south_west
  }

  stack = []

  rules.each do |rule|
    direction, a, b = rule.split

    if pair_map.key?(direction)
      dir = pair_map[direction]
      if (dir == :north_east && stack.last == :south_west) ||
         (dir == :north_west && stack.last == :south_east) ||
         (dir == :south_east && stack.last == :north_west) ||
         (dir == :south_west && stack.last == :north_east)
        return false
      end
    elsif direction_map.key?(direction)
      opposite_direction = direction_map[direction]
      if stack.last == opposite_direction
        return false
      end
    end

    stack << dir
  end

  true
end
```